<h2 id="introduction">Introduction</h2>
<p>This document describes the assay for transposase-accessible chromatin using sequencing (ATAC-seq) genome browser tracks for the project &#8216;Neuronal Brain Region-Specific DNA Methylation and Chromatin Accessibility are Associated with Neuropsychiatric Disease Heritability&#8217;. For an overview of the project, please see the <a href="https://s3.us-east-2.amazonaws.com/brainepigenome/hg19/docs/README.html">README</a>.</p>
<h2 id="mapping-and-quality-control-of-atac-seq-reads">Mapping and quality control of ATAC-seq reads</h2>
<p>We trimmed reads of their adapter sequences using <a href="https://github.com/lh3/trimadap/archive/0.1.zip"><strong>trimadap</strong> (v0.1)</a> with the following parameters: <code>trimadap-mt -3 CTGTCTCTTATACACATCTCCGAGCCCACGAGA ${READ1}; trimadap-mt -3 CTGTCTCTTATACACATCTGACGCTGCCGACGA ${READ2}</code></p>
<p>We then aligned these trimmed reads to the hg19 build of the human genome (including autosomes, sex chromosomes, mitochondrial sequence, unplaced sequence, and unlocalized sequence) using <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml"><strong>Bowtie2</strong> (v2.2.5)</a> with alignment parameters: <code>bowtie2 -X 2000 --local --dovetail</code>. Potential PCR duplicate reads were marked using MarkDuplicates from the <a href="http://broadinstitute.github.io/picard/">Picard library (v2.2.1)</a>.</p>
<h2 id="identifying-differentially-accessible-regions-dars">Identifying differentially accessible regions (DARs)</h2>
<p>Peaks were called in each condition (NAcc_pos, NAcc_neg, BA9_pos, and BA9_neg) using <a href="https://github.com/taoliu/MACS">MACS (v2.1.0)</a>. Specifically, data from each condition were combined into a metasample formed by using all non-duplicate-marked reads with a mapping quality &gt; 30 and then processed using: <code>macs2 callpeaks --nomodel --nolambda --call-summits -t ${BAMS[@]}</code>. We took the &#8216;narrowPeaks&#8217; produced by MACS and filtered out those regions overlapping the <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/">ENCODE mappability consensus blacklist regions</a> and the <a href="https://sites.google.com/site/atacseqpublic/atac-seq-analysis-methods/mitochondrialblacklists-1">blacklist for ATAC-seq created by Buenrostro et al (access requires signing in with a Google Account)</a>. We took this filtered list as our condition-specific sets of open chromatin regions (OCRs).</p>
<p>To perform the differential analysis, we first took the union of condition-specific OCRs on the autosomes to construct an &#8216;overall&#8217; set of OCRs. This &#8216;overall&#8217; set of OCRs contained 853,053 regions (630 Mb). For each sample, we then counted the number of fragments (fragment = start of read1 to end of read2) overlapping each of the &#8216;overall&#8217; OCRs using the <code>summarizeOverlaps()</code> function in the <a href="https://bioconductor.org/packages/GenomicAlignments/"><strong>GenomicAlignments</strong> R/Bioconductor package (v1.10.0)</a>. Specifically, we only counted those fragments where both reads had a mapping-quality score &gt; 30, reads not marked as potential PCR duplicates, and those where any part of the fragment overlapped exactly one peak.</p>
<p>We then analyzed these data using the <a href="https://www.bioconductor.org/packages/limma/"><strong>voom</strong> method</a>, originally designed for differential expression analysis of RNA-seq data17. Briefly, the read counts were transformed to counts per million (cpm) and only those 283,812 / 853,053 peaks with at least 1 cpm for at least 5 samples (the size of the smallest group of samples) were retained. We normalized these counts using TMM, then used <a href="https://www.bioconductor.org/packages/edgeR/"><strong>edgeR</strong> (v3.16.5)</a> and <a href="https://www.bioconductor.org/packages/limma/"><strong>limma</strong> (v3.30.7)</a> to transform these counts to log2-cpm, estimate the mean-variance relationship, and compute appropriate observation-level weights ready for linear modelling.</p>
<p>In our design matrix, we blocked on donor (donor1, &#8230;, donor6) and included a term for each group (e.g., BA9_neg for NeuN- cells from BA9, BA9_pos for NeuN+ cells from BA9, etc.). We ran surrogate variable analysis21 using the <a href="https://www.bioconductor.org/packages/sva/"><strong>sva</strong> (v3.22.0) R/Bioconductor package</a> and identified 4 surrogate variables, one of which correlated with the date on which these samples were flow-sorted. We ultimately decided to include all 4 surrogate variables in the linear model. Using the empirical Bayes shrinkage method implemented in <a href="https://www.bioconductor.org/packages/limma/"><strong>limma</strong></a>, we tested for differential accessibility of peaks in three comparisons:</p>
<ol type="1">
<li>NAcc vs.&#160;BA9 in NeuN+ cells</li>
<li>NAcc vs.&#160;BA9 in NeuN- cells</li>
<li>NeuN+ cells vs.&#160;NeuN- cells</li>
</ol>
<p>For an ATAC-seq peak to be called a differentially accessible region (DAR), it had to have a Benjamini-Hochberg adjusted P-value &lt; 0.05.</p>
<h2 id="credits">Credits</h2>
<p>Questions should be directed to <a href="mailto:peter.hickey@gmail.com">Peter Hickey</a>.</p>
