<h2 id="introduction">Introduction</h2>
<p>This document describes the RNA-seq genome browser tracks for the project &#8216;Neuronal Brain Region-Specific DNA Methylation and Chromatin Accessibility are Associated with Neuropsychiatric Disease Heritability&#8217;. For an overview of the project, please see the <a href="https://s3.us-east-2.amazonaws.com/brainepigenome/hg19/docs/README.html">README</a>.</p>
<h2 id="mapping-and-quality-control-of-rna-seq-reads">Mapping and quality control of RNA-seq reads</h2>
<p>We trimmed the first 3 bp of read1, which were derived from template switching oligos and not the cDNA of interest, using <a href="https://github.com/lh3/seqtk"><strong>seqtk</strong> (v1.2-r94)</a> with the following parameters: <code>seqtk trimfq -b 3 ${READ1}</code>. We then quasi-mapped these trimmed reads to a FASTA file of protein-coding and lncRNA genes from <a href="http://www.gencodegenes.org/releases/19.html">GENCODE v19</a> and performed transcript-level quantification using <a href="https://github.com/COMBINE-lab/salmon"><strong>Salmon</strong> (v0.7.2)</a>.</p>
<h2 id="identifying-differentially-expressed-genes-degs">Identifying differentially expressed genes (DEGs)</h2>
<p>We used <a href="https://www.bioconductor.org/packages/tximport/"><strong>tximport</strong> (v1.2.0)</a> to compute normalized gene-level counts from the transcript-level abundance estimates (scaling these using the average transcript length over samples and the library size). Only autosomal genes with at least 1 cpm in at least 4 libraries (the size of the smallest group of samples) were retained for downstream analysis (24,161 / 33,351 genes). We normalized these counts using TMM18 then used <a href="https://www.bioconductor.org/packages/edgeR/"><strong>edgeR</strong> (v3.16.5)</a> and <a href="https://www.bioconductor.org/packages/limma/"><strong>limma</strong> (v3.30.7)</a> to transform these counts to log2-cpm, estimate the mean-variance relationship, and compute appropriate observation-level weights ready for linear modelling.</p>
<p>In our design matrix, we blocked on donor (donor1, &#8230;, donor6) and included a term for each group (e.g., BA9_neg for NeuN- cells from BA9, BA9_pos for NeuN+ cells from BA9, etc.). We ran surrogate variable analysis21 using the <a href="https://www.bioconductor.org/packages/sva/"><strong>sva</strong> (v3.22.0) R/Bioconductor package</a> and identified 5 surrogate variables, some of which correlated with the date on which these samples were flow-sorted. We ultimately decided to include all 5 surrogate variables in the linear model. Using the empirical Bayes shrinkage method implemented in <a href="https://www.bioconductor.org/packages/limma/"><strong>limma</strong></a>, we tested for differential expression of genes in three comparisons:</p>
<ol type="1">
<li>NAcc vs.&#160;BA9 in NeuN+ cells</li>
<li>NAcc vs.&#160;BA9 in NeuN- cells</li>
<li>NeuN+ cells vs.&#160;NeuN- cells</li>
</ol>
<p>For a gene to be called a differentially expressed gene (DEG), it had to have a Benjamini-Hochberg adjusted P-value &lt; 0.05 with no minimum log2 fold change cutoff.</p>
<h2 id="credits">Credits</h2>
<p>Questions should be directed to <a href="mailto:peter.hickey@gmail.com">Peter Hickey</a>.</p>
